### PR TITLE
LF: Test scala interface type checking

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description
- [ ] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by remyhaemmerle-da** [comment]
I suggest to use "Hide whitespace" option to review this PR.

File: daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala [hunk 1]
```diff
@@ -744,15 +759,25 @@ class TypingSpec extends AnyWordSpec with TableDrivenPropertyChecks with Matcher
...
+            case EUnknownDefinition(
+                  _,
+                  LookupError(Reference.Interface(_), Reference.Interface(_)),
+                ) =>
+          },
+        E"Λ (σ : ⋆). λ (e: σ) → ⸨ create_by_interface @Mod:I e ⸩" -> //
+          { case _: ETypeMismatch => }, */
         // UpdExercise
-        E"λ (e₂: List Party) (e₃: Int64) → ⸨ exercise @Mod:U Ch nothing e₂ e₃ ⸩" -> //
+        E"λ (e₁: ContractId Mod:U) (e₂: List Party) (e₃: Int64) → ⸨ exercise @Mod:U Ch e₁ e₂ e₃ ⸩" -> //
```
Comment (cocreature) [inline daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala:773]: I’m confused, isn’t there one argument too much here? It seems like exercise should only have the cid & choice arg but we have e1, e2, e3

File: daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala [hunk 2]
```diff
@@ -767,20 +792,56 @@ class TypingSpec extends AnyWordSpec with TableDrivenPropertyChecks with Matcher
...
           },
-        E"""⸨ fetch_by_key @Mod:T "Bob" ⸩""" -> //
+        E"λ (e₁: ContractId Mod:I) (e₂: List Party) (e₃: Int64) → ⸨ exercise_by_interface @Mod:I Not e₁ e₂ e₃ ⸩" -> //
+          {
+            case EUnknownDefinition(
+                  _,
+                  LookupError(Reference.TemplateChoice(_, _), Reference.InterfaceChoice(_, _)),
+                ) =>
+          },
+        E"Λ (σ : ⋆).λ (e₁: ContractId Mod:I) (e₂: List Party) (e₃: σ) → ⸨ exercise_by_interface @Mod:T ChIface e₁ e₂ e₃ ⸩" -> //
```
Comment (cocreature) [inline daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala:822]: same question here, it looks like you’re trying to pass actors for some reason.

File: daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala [hunk 3]
```diff
@@ -767,20 +792,56 @@ class TypingSpec extends AnyWordSpec with TableDrivenPropertyChecks with Matcher
...
           { case _: ETypeMismatch => },
-        E"""⸨ lookup_by_key @Mod:T "Bob" ⸩""" -> //
+        E"Λ (σ : ⋆).λ (e₁: ContractId Mod:I) (e₂: List σ) (e₃: Int64) → ⸨ exercise_by_interface @Mod:T ChIface e₁ e₂ e₃ ⸩" -> //
+          { case _: ETypeMismatch => },
+        E"Λ (σ : ⋆).λ (e₁: ContractId Mod:I) (e₂: σ) (e₃: Int64) → ⸨ exercise_by_interface @Mod:T ChIface e₁ e₂ e₃ ⸩" -> //
+          { case _: ETypeMismatch => },
+        E"Λ (σ : ⋆).λ (e₁: ContractId σ) (e₂: List Party) (e₃: Int64) → ⸨ exercise_by_interface @Mod:T ChIface e₁ e₂ e₃ ⸩" -> //
           { case _: ETypeMismatch => },
+        // This verifies that interface choice cannot be exercise by template
```
Comment (cocreature) [inline daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala:830]:
```suggestion
        // This verifies that interface choice cannot be exercised by template
```

**Review by cocreature — APPROVED** [review]
Nice! Two main comments:
1. I don’t understand the tests for actors.
2. I don’t understand why we don’t allow that a template with a key implements an interface.

File: daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala [hunk 4]
```diff
@@ -417,6 +431,7 @@ class TypingSpec extends AnyWordSpec with TableDrivenPropertyChecks with Matcher
 
       val testCases = Table[Expr, PartialFunction[ValidationError, _]](
         "non-well formed expression" -> "error",
+        /*
```
Comment (sofiafaro-da) [inline daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala:434]: Accidental?

**Review by sofiafaro-da — APPROVED** [review]
Thank you!

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala [hunk 5]
```diff
@@ -451,17 +451,17 @@ private[validation] object Typing {
         ()
       }
       implementations.values.foreach(env.checkIfaceImplementation(tplName, _))
+      if (mbKey.isDefined && implementations.nonEmpty)
+        throw ETemplateWithKeyAndInterface(ctx, tplName)
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala]: Reverted.

File: daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala (hunk 4 above)
Comment (remyhaemmerle-da) [reply daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala]: Yes.

File: daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala (hunk 1 above)
Comment (remyhaemmerle-da) [reply daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala]: Good catch.

File: daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala (hunk 2 above)
Comment (remyhaemmerle-da) [reply daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala]: fixed

File: daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala (hunk 3 above)
Comment (remyhaemmerle-da) [reply daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala:823]: fixed

**Comment by remyhaemmerle-da** [comment]
> Nice! Two main comments:
> 1. I don’t understand the tests for actors.

I fixed the exercise tests.

> 2. I don’t understand why we don’t allow that a template with a key implements an interface.

I dropped the additional check.

**Pull request merged**